File: 17-Binary-Tree/java/12-MaximumWidthOfBinaryTree.java
class Solution {
    private int maxWidth;
    private List<Long> leftmostIndices;

    public int widthOfBinaryTree(TreeNode root) {
        maxWidth = 0;
        leftmostIndices = new ArrayList<>();
        dfs(root, 0, 0L);
        return maxWidth;
    }

    private void dfs(TreeNode node, int depth, long index) {
        if (node == null) return;

        if (depth == leftmostIndices.size()) {
            leftmostIndices.add(index);
        }

        long width = index - leftmostIndices.get(depth) + 1;
        maxWidth = Math.max(maxWidth, (int) width);

        long normalizedIndex = index - leftmostIndices.get(depth);
        dfs(node.left, depth + 1, 2 * normalizedIndex);
        dfs(node.right, depth + 1, 2 * normalizedIndex + 1);
    }
}

File: 17-Binary-Tree/python/12_maximum_width_of_binary_tree.py
class Solution:
    def widthOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        max_width = 0
        leftmost_indices = []

        def dfs(node, depth, index):
            nonlocal max_width
            if not node:
                return

            if depth == len(leftmost_indices):
                leftmost_indices.append(index)

            width = index - leftmost_indices[depth] + 1
            max_width = max(max_width, width)

            normalized_index = index - leftmost_indices[depth]
            dfs(node.left, depth + 1, 2 * normalized_index)
            dfs(node.right, depth + 1, 2 * normalized_index + 1)

        dfs(root, 0, 0)
        return max_width
